fdmi/plugins/fdmi_app

```diff
@@ -60,19 +62,119 @@ MAX_KEYSTRS = 10240
 # is communicating.
 processes = []
 
+FDMI_producer = None
+ack_thread = None
+lock = threading.Lock()
+
+
+def send_ack_to_plugin(banner, p, rec_id):
+    print(banner + " ACK " + rec_id, flush = True)
+    ri = bytes('{}\n'.format(rec_id), encoding = "utf-8")
+    lock.acquire()
+    p.stdin.write(ri)
+    p.stdin.flush()
+    lock.release()
+
+
+
+class Acker (threading.Thread):
+    def __init__(self, threadID, name, args):
+        threading.Thread.__init__(self)
+        self.threadID = threadID
+        self.name = name
+        self.args = args
+        self.consumer = None
+        self.done = False
+
+    def run(self):
+        if self.args.kafka_server is not None:
+            print("Starting: " + self.name, flush = True)
+            self.consume_and_ack()
+
+    def terminate(self):
+        if self.args.kafka_server is not None:
+            self.done = True
+            print("Flagging " + self.name + " Done", flush = True)
+
+    def consume_and_ack(self):
```

Comment:
When it consumes, then it sends ack? My understanding was that we send ack once written to Kafka and consumer is the replication manager. When is this KafkaConsume run?

Reply:
This is an internal consumer. It has nothing related to Replication Manager.

Sending records to Kafka is asynchronous.
If we wait for the result of a sending, it becomes SYNCHRONOUS way. 
In my testing in a single VM, we can send ~1000 messages per second asynchronously.
But we can send only ~100 messages per second if we send and wait (SYNCHRONOUS way).
So, I introduce an internal consumer. When this internal consumer gets some message, it means these messages are persistent. (Replication manager can get these messages too. They are independent).

Reply:
I think producer does batching in the background so async producer will look much faster. 
 
Can we use callback in producer instead:  
https://docs.confluent.io/clients-confluent-kafka-python/current/overview.html#asynchronous-writes

producer(topic, key, value, callback_func)
callback_func(err, msg)
   if no error, send_ack_to_plugin(msg)

Reply:
I will use callback mechanism.